File: searching/binarySearchRec.cpp
/*
goal : to search an element in sorted array
time - O(logn)
space - O(1)
*/

#include<bits/stdc++.h>
#include<iostream>
using namespace std;

int binarySearch(int arr[], int l , int h , int key)
{
    if (h == 1)
    {   
        if (arr[l] ==  key)
        {
            return l;
        }
        else
        {
            return -1;
        }
        
    }
    
    while (l<=h)
    {
        int mid = (l+h)/2;

        if (arr[mid] == key)
        {
            return mid;
        }
        else if (arr[mid] > key)
        {
            return binarySearch(arr,l,mid-1, key);
        }
        else
        {
            return binarySearch(arr,mid+1, h, key);
        }
        
    }
    return -1;
    
}
int main()
{
    int arr[] = {4,32,54,67,75,89};
    int n = sizeof(arr)/ sizeof(arr[0]);

    cout << binarySearch(arr,0,n,54);
    return 0;
}

File: searching/binarySearchRec.java
public class binarySearchRec {

    public static int search(int[] a, int l, int h, int target) {

        while (l <= h) {

            int mid = (l + h) / 2;

            if (l == h) {
                if (a[l] == target) {
                    return l;
                } else {
                    return -1;
                }

            } else {
                if (target == a[mid]) {
                    return mid;
                } else if (target < a[mid]) {
                    return search(a, l, mid - 1, target);
                } else {
                    
                    return search(a, mid + 1, h, target);
                }
            }

        }
        return -1;

    }

    public static void main(String[] args) {

        int[] a = {3,5,65,23,121,23};
        int n = a.length;
        int ans = search(a, 0, n-1, 121);
        System.out.println(ans);

    }

}
